### PR TITLE
CORE-9787: Enable local developers to pull from remote build cache

### PR DESCRIPTION
**Pull request by knguyen-r3**
* Enable usage of build cache locally
* Implemented new identity access base using gradleEnterprise.buildCache

### CODE REVIEW

**Comment by chrisr3** [comment]
Why on Earth would we want to enable the build cache locally? IIRC, the last time we had that turned on, people were trying to debug their JUnit tests ,and wondering why they were completing immediately without stopping at any breakpoints?!

**Comment by ronanbrowne** [comment]
@chrisr3 , thanks for flagging.

developer productivity is becoming more of an issue as runtime os grows to more than 300 submodels, 

Maybe before my time on the test issue you mentioned, as this is the first time we have configured grades remote cache to true for local use.
point taken, However, we should leverage the cache for compilation optimization / image building etc

as a workaround, we can do something like this to avoid test caching locally based on remote outputs

```
tasks.withType(Test).configureEach {
 if (!System.getenv().containsKey("JENKINS_URL")) {  // not a CI env so local user
     outputs.upToDateWhen { false }
  }
}
```

**Comment by driessamyn** [comment]
> @chrisr3 , thanks for flagging.
> 
> developer productivity is becoming more of an issue as runtime os grows to more than 300 submodels,
> 
> Maybe before my time on the test issue you mentioned, as this is the first time we have configured grades remote cache to true for local use. point taken, However, we should leverage the cache for compilation optimization / image building etc
> 
> as a workaround, we can do something like this to avoid test caching locally based on remote outputs
> 
> ```
> tasks.withType(Test).configureEach {
>  if (!System.getenv().containsKey("JENKINS_URL")) {  // not a CI env so local user
>      outputs.upToDateWhen { false }
>   }
> }
> ```

Or we can teach people how to run a build without the cache.
Personally, I would appreciate not having to sit through the 13 minutes of "unit" (they're so no unit tests) tests, 99% of which aren't impacted by my change.

**Comment by ronanbrowne** [comment]
@chrisr3 / @driessamyn  I reached out to Gradle on this via their community slack channel for Gradle Ent support and got the following response.

TLDR: delegate to gradle let it decide, IntelliJ  will always choose to re-run a test if you explicitly tell it to even if cach is up-to-date

"IntelliJ usually also cares for the test always being rerun using an init script if you "run tests" even if they are delegated to Gradle, because IntelliJ has the opinion, if a user asks to run the tests, they should run and not be skipped.

I personally would not recommend disabling the cache for such a reason even if they were up-to-date or cached when run through IntelliJ properly. Even if one wonders why no breakpoints are hit an the test finished immediately, that should only happen once with confusion and then it should be clear what happened and why the next time.

But if you must, there are multiple possibilities:
- disable caching for test tasks
- disable caching for test tasks when run through IntelliJ
- disable caching for test tasks if some Gradle property is set so that the ones insisting can have the behaviour they want with the flick of a switch
- enable caching and tell people that insist to either disable the caching completely by setting in their Gradle user home, then it is also for all projects they work on, not only a specific one
- enable caching and tell people to use an init script in their Gradle user home to do whatever they want, like for example disabling caching on test tasks."


I am in favour here of going with gradles recommendation activating and giving someone the option to disable caching should they _really_ need to. Which can be done via `--no-build-cache` or `org.gradle.caching=false`

File: settings.gradle [hunk 1]
```diff
@@ -484,35 +484,30 @@ if (compositeBuild.toBoolean() && file(cordaCliHostLocation).exists()) {
...
-            }
+        remote(gradleEnterprise.buildCache) {
             // For the remote build cache we will populate cache only from Jenkins, all machines can pull from cache however.
             if (System.getenv().containsKey("JENKINS_URL")) {
                 push = true
                 enabled = true
             } else {
                 push = false
-                enabled = false
+                enabled = true
```
Comment (ronanbrowne) [inline settings.gradle]: We need to make sure there is no issue if a dev doesn't have a valid API key set

```suggestion
                enabled = apiKey?.trim() ? true : false
```

**Review by ronanbrowne — CHANGES_REQUESTED** [review]
see comment

**Review by yarivus — APPROVED** [review]
LGTM

**Review by ronanbrowne — APPROVED** [review]
LGTM
